### ISSUE TITLE
DEP0040  `punycode` deprecation warning with Node.js 22.x and later

### ISSUE THREAD

**Issue by MikeMcC399** [tcort/markdown-link-check#365]
## Current behavior

Running `markdown-link-check` under Node.js `22.x` (or `23.x`) results in a `punycode` deprecation warning, similar to

```text
(node:4374) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
```

## Expected behavior

`markdown-link-check` should not show any deprecation warning running under Node.js `22.x`.

According to the [Node.js release schedule](https://github.com/nodejs/release#release-schedule) version Node.js `22.x` moved to Active LTS status on Oct 29, 2024.

## Steps to reproduce

| Component           | Version       |
| ------------------- | ------------- |
| Ubuntu              | `24.04.1` LTS |
| Node.js             | ~~`22.10.0`~~  `22.11.0`     |
| markdown-link-check | `3.12.2`      |

```shell
mkdir markdown-link-check-test
cd markdown-link-check-test
npm init -y
npm install markdown-link-check@latest
touch test.md
npx markdown-link-check test.md
```

## Logs

```text
$ npx markdown-link-check test.md
(node:4374) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)

FILE: test.md
  No hyperlinks found!

  0 links checked.
```

and with trace enabled:

```text
$ NODE_OPTIONS='--trace-deprecation' npx markdown-link-check test.md
(node:4405) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
    at node:punycode:3:9
    at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/realm:399:7)
    at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/realm:338:10)
    at loadBuiltinModule (node:internal/modules/helpers:114:7)
    at Function._load (node:internal/modules/cjs/loader:1100:17)
    at TracingChannel.traceSync (node:diagnostics_channel:315:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:218:24)
    at Module.require (node:internal/modules/cjs/loader:1340:12)
    at require (node:internal/modules/helpers:141:16)
    at Object.<anonymous> (/home/mike/github/markdown-link-check-test/node_modules/isemail/lib/index.js:5:18)

FILE: test.md
  No hyperlinks found!

  0 links checked.
```

```text
$ npm ls isemail
markdown-link-check-test@1.0.0
└─┬ markdown-link-check@3.12.2
  └─┬ link-check@5.3.0
    └── isemail@3.2.0
```

## Comment

In [isemail](https://github.com/skeggse/isemail) the  `punycode` deprecation issue appears to have been fixed in the `master` branch and not released. The last released version was `isemail@3.2.0` released on Oct 18, 2018 and now https://github.com/skeggse/isemail/issues is archived since Oct 2, 2024, meaning that the deprecation won't ever be released and that the repo is unsupported.

`link-check@5.3.0` is the latest version. I guess that https://github.com/tcort/link-check would need to look for an alternative to [isemmail](https://github.com/skeggse/isemail) to resolve this issue.

**Comment by MikeMcC399** [comment]
Under Node.js `23.0.0` there is an additional experimental warning, which I didn't look at so far. The trace does not provide any additional information, and so I would wait for a follow-on version of Node.js `23.x` before wanting to dig deeper:

```text
$ npx markdown-link-check test.md
(node:4986) ExperimentalWarning: Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:4998) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
```

**Comment by MikeMcC399** [comment]
- Issue https://github.com/tcort/link-check/issues/92 submitted for dependency [link-check](https://github.com/tcort/link-check)

**Comment by danielhoherd** [comment]
FWIW this error has finally gotten to a place where I am unable to get pre-commit to run. I have had to remove the pre-commit hook from my repo in order to prevent this failure from causing pre-commit to fail entirely.

```
(node:43379) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
/Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/node_modules/async/dist/async.js:325
            if (fn === null) throw new Error("Callback was already called.");
                             ^

Error: Callback was already called.
    at /Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/node_modules/async/dist/async.js:325:36
    at /Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/node_modules/async/dist/async.js:249:17
    at /Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/index.js:168:13
    at /Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/node_modules/link-check/lib/proto/http.js:118:21
    at done (/Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/node_modules/needle/lib/needle.js:474:14)
    at PassThrough.<anonymous> (/Users/danielh/.cache/pre-commit/reporwc5do1t/node_env-system/lib/node_modules/markdown-link-check/node_modules/needle/lib/needle.js:734:9)
    at PassThrough.emit (node:events:519:35)
    at endReadableNT (node:internal/streams/readable:1696:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:90:21)

Node.js v23.1.0
```

**Comment by tcort** [comment]
I will look at pull requests tomorrow and roll a new release.

**Comment by MikeMcC399** [comment]
@danielhoherd 

Sounds like that would be a different issue. In other situations the `DeprecationWarning` does not cause a failure.

**Comment by danielhoherd** [comment]
@MikeMcC399 wow, you're right. i was able to work around that stack trace by doing `pre-commit uninstall ; pre-commit install ;` My bad! Thanks for the nudge!

**Comment by MikeMcC399** [comment]
This issue is resolved in [markdown-link-check@3.13.5](https://github.com/tcort/markdown-link-check/releases/tag/v3.13.5)
